File: 2023 ICPC Asia Chennai Onsite Contest/Basic_Blooms_v1.3.py
import heapq
import sys

MOD = 998244353
N = 1_000_000

nums = [0] * N
prefix_sums = [0] * (N + 1)

class Num:
    def __init__(self, base):
        self.value = 1
        self.base_value = 1
        self.mod_value = 1
        self.mod_base_value = 1
        self.base = base
        self.digit = 1

    def increment(self):
        self.digit += 1
        self.value += self.base_value
        self.mod_value = (self.mod_value + self.mod_base_value) % MOD
        if self.digit == self.base:
            self.digit = 1
            self.value += 1
            self.base_value = self.value
            self.mod_value = (self.mod_value + 1) % MOD
            self.mod_base_value = self.mod_value

    def __gt__(self, other):
        return self.value > other.value

def precompute():
    nums = [0] * N
    prefix_sums = [0] * (N + 1)
    prev_num = 0
    pq = []
    for base in range(2, 17):
        heapq.heappush(pq, Num(base))

    i = 0
    while i < N:
        num = heapq.heappop(pq)
        if num.value > prev_num:
            nums[i] = num.mod_value
            prev_num = num.value
            i += 1
        num.increment()
        heapq.heappush(pq, num)

    for i in range(1, N + 1):
        prefix_sums[i] = (prefix_sums[i - 1] + nums[i - 1]) % MOD
        
    return prefix_sums

def main():
    prefix_sums = precompute()
    t = int(sys.stdin.readline())
    for _ in range(t):
        l, r = map(int, sys.stdin.readline().split())
        print((prefix_sums[r] - prefix_sums[l - 1] + MOD) % MOD)

if __name__ == "__main__":
    main()

File: 2023 ICPC Asia Chennai Onsite Contest/Basic_Blooms_v1.3.cpp
#include <bits/stdc++.h>
#include <gmpxx.h> // Compile with -lgmpxx -lgmp

using namespace std;

using ll = mpz_class;
constexpr int MOD = 998244353;
constexpr int N = 1'000'000;

array<int, N> nums;
array<int, N + 1> prefix_sums;

struct Num {
    ll value;
    ll base_value;
    int mod_value;
    int mod_base_value;
    int base;
    int digit;

    Num(int base): base(base), digit(1), value(1), base_value(1), mod_value(1), mod_base_value(1) {}

    Num& operator++() {
        digit++;
        value += base_value;
        (mod_value += mod_base_value) %= MOD;
        if (digit == base) {
            digit = 1;
            value += 1;
            base_value = value;
            (mod_value += 1) %= MOD;
            mod_base_value = mod_value;
        }

        return *this;
    }

    bool operator>(const Num& other) const {
        return value > other.value;
    }
};

void precompute() {
    ll prev_num = 0;
    priority_queue<Num, vector<Num>, greater<Num>> pq;
    for (int base = 2; base <= 16; ++base) {
        pq.emplace(base);
    }

    for (int i = 0; i < N; ++i) {
        auto num = pq.top(); pq.pop();
        if(num.value > prev_num) {
            nums[i] = num.mod_value;
            prev_num = num.value;
        } else {
            --i;
        }
        ++num;
        pq.push(num);
    }

    prefix_sums[0] = 0;
    for (int i = 0; i < N; ++i) {
        prefix_sums[i + 1] = (prefix_sums[i] + nums[i]) % MOD;
    }
}

int main() {
    precompute();
    int t;
    cin >> t;
    while (t--) {
        int l, r; cin >> l >> r;
        cout << (prefix_sums[r] - prefix_sums[l - 1] + MOD) % MOD << '\n';
    }
    return 0;
}